### PR TITLE
Fix exp2 derivatives overflowing to inf near the float64 max

### LINKED ISSUE #192408: torch.exp2 second derivative overflows to inf at 1024.0 although the float64 value is finite
### 🐛 Describe the bug

`torch.autograd` returns `inf` for the second derivative of `torch.exp2` at a `float64` input where the mathematically correct second derivative is finite and representable in `float64`.

The second derivative of `exp2(x)` is `(ln(2) ** 2) * 2^x`. For `x = 1024.0`, the true second derivative is approximately `8.637070847446594e+307`, which is below the maximum finite `float64` value. However, PyTorch returns `inf`.

This issue reproduces on CPU, so CUDA availability does not affect the result.

```python
import torch

torch.set_default_dtype(torch.float64)

x = torch.tensor(1024.0, requires_grad=True)

y = torch.exp2(x)

g1, = torch.autograd.grad(y, x, create_graph=True)
g2, = torch.autograd.grad(g1, x)

actual = g2
expected = torch.tensor(8.637070847446594e+307, dtype=torch.float64)

print("actual:")
print(actual)
print("expected:")
print(expected)
```

### Actual vs expected result

```text
actual:
tensor(inf)

expected:
tensor(8.6371e+307)
```

The discrepancy occurs in the second-order backward pass. Although the forward value `torch.exp2(1024.0)` overflows, the mathematical second derivative `(ln(2) ** 2) * 2^1024` is still finite in `float64`, so the second derivative should not be `inf`.

### Versions

```text
PyTorch version: 2.12.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
Python version: 3.13.13 | packaged by Anaconda, Inc.
Python platform: Linux x86_64

Is CUDA available: False
GPU models:
  GPU 0: NVIDIA RTX 6000 Ada Generation
  GPU 1: NVIDIA RTX 6000 Ada Generation
Nvidia driver version: 570.211.01

CPU:
  Model name: AMD Ryzen Threadripper PRO 7985WX 64-Cores
  CPU(s): 128

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
[pip3] triton==3.7.0
```

### PR DESCRIPTION
**Pull request by Aayaan-Sahu**
### Changes
torch's exp2 derivative was computed using: `grad * result.conj() * M_LN2` in derivatives.yaml

2**x overflows float64 at x=1024, but the first derivative stays finite until ~1024.53 and the second derivative stays finite until ~1025.06. Currently, since inf is at x = 1024, multiplying by M_LN2 propagates the inf.

The fix is:
`self: grad * exp2(self - 1).conj() * (2 * M_LN2)`

Fixes #192408

### Test Plan:

  New regression test asserting both derivatives at x = 1024. The expected
  values are built with math.ldexp so that computing the reference does not
  itself overflow:

  ```
  python test/test_autograd.py -k exp2
  ```

  Existing exp2 coverage, including OpInfo gradcheck/gradgradcheck:

  ```
  python test/test_ops_gradients.py -k exp2
  python test/test_ops_fwd_gradients.py -k exp2
  python test/test_unary_ufuncs.py -k exp2
  python test/test_ops.py -k exp2
  ```

### CODE REVIEW

**Comment by Skylion007** [comment]
Doesn't this hurt really small derivatives which are more likely?

**Comment by Aayaan-Sahu** [comment]
This would hurt really small values, yes. I think that it may not be worth to protect upper bounds because a really large input to exp2 probably means that gradients will explode anyways.
